pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,213 @@
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_acquire_failed_total",
+			Help: "Total number of failed attempts to acquire concurrency slots",
+		}),
+		ReleaseTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_release_total",
+			Help: "Total number of released concurrency slots",
+		}),
+	}
+}
+
+// MultiTenantConcurrencyController is a concurrency controller that limits the number of concurrent rule evaluations both global and per tenant.
+type MultiTenantConcurrencyController struct {
+	logger               log.Logger
+	limits               RulesLimits
+	maxGlobalConcurrency int64
+	metrics              *MultiTenantConcurrencyControllerMetrics
+
+	globalConcurrency semaphore.Weighted
+	mtx               sync.Mutex
+	tenantConcurrency map[string]*DynamicSemaphore
+}
+
+// NewMultiTenantConcurrencyController creates a new MultiTenantConcurrencyController.
+func NewMultiTenantConcurrencyController(logger log.Logger, maxGlobalConcurrency int64, limits RulesLimits, reg prometheus.Registerer) *MultiTenantConcurrencyController {
+	return &MultiTenantConcurrencyController{
+		metrics:              newMultiTenantConcurrencyControllerMetrics(reg),
+		logger:               log.With(logger, "component", "ruler.MultiTenantConcurrencyController"),
+		maxGlobalConcurrency: maxGlobalConcurrency,
+		limits:               limits,
+		globalConcurrency:    *semaphore.NewWeighted(maxGlobalConcurrency),
+		tenantConcurrency:    make(map[string]*DynamicSemaphore),
+	}
+}
+
+// Done releases a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Done(ctx context.Context) {
+	c.globalConcurrency.Release(1)
+	c.metrics.CurrentConcurrency.Dec()
+	c.metrics.ReleaseTotal.Inc()
+
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error extracting org id from context while releasing a slot", "err", err)
+		return
+	}
+
+	c.mtx.Lock()
+	tc, ok := c.tenantConcurrency[userID]
+	c.mtx.Unlock()
+
+	if ok {
+		tc.Release()
+	} else {
+		c.logger.Log("msg", "tenant concurrency controller: tenant not found", "tenant", userID)
+	}
+}
+
+// Allow tries to acquire a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Allow(ctx context.Context, group *rules.Group, rule rules.Rule) bool {
+	c.metrics.AcquireTotal.Inc()
+	// To allow a rule to be executed concurrently, we need 3 conditions:
+	// 1. The rule group must be at risk of missing its evaluation.
+	// 2. The rule must not have any rules that depend on it.
+	// 3. The rule itself must not depend on any other rules.
+	if !isGroupAtRisk(group) {
+		return false
+	}
+
+	if !isRuleIndependent(rule) {
+		return false
+	}
+
+	// Next, try to acquire a global concurrency slot.
+	if !c.globalConcurrency.TryAcquire(1) {
+		c.metrics.AcquireFailedTotal.Inc()
+		return false
+	}
+	c.metrics.CurrentConcurrency.Inc()
```

Comment:
We increase this metric and then revert if an error occurred. I would keep it simple and just increase this metric right before we `return true`

Reply:
Discussed this offline. Because we're measuring "global slots" and not per tenant slots, the current approach is correct. As you said, it's a very subtle distinction as it really depends on the scrape interval but let me know if you change your mind and I'll change it.